adonis-typings/drive.ts

```diff
@@ -47,10 +50,99 @@ declare module '@ioc:Adonis/Core/Drive' {
     etag?: string
   }
 
+  /**
+   * List item returned by the drive drivers
+   */
+  export interface DriveListItem<T = any> {
+    /**
+     * Location of list item on disk which can be used in driver methods
+     */
+    location: string
+
+    /**
+     * Flag to know if item represents file or directory
+     */
+    isFile: boolean
+
+    /**
+     * Original list item returned from underlying driver
+     */
+    original: T
+  }
+
+  /**
+   * Shape of the directory listing async iterable returned from list allowing to transform listing.
+   * This can be iterated directly using for-await-of loop or it can be converted to array using toArray().
+   */
+  export interface DirectoryListingContract<Driver extends DriverContract, T>
+    extends AsyncIterable<T> {
+    /**
+     * Reference to the driver for which the listing was created.
+     */
+    driver: Driver
+
+    /**
+     * Filter generated items of listing with the given predicate function.
+     */
+    filter(
+      predicate: (item: T, index: number, driver: Driver) => Promise<boolean> | boolean
+    ): DirectoryListingContract<Driver, T>
+
+    /**
+     * Transform generated items of listing with the given mapper function.
+     */
+    map<M>(
+      mapper: (item: T, index: number, driver: Driver) => Promise<M> | M
+    ): DirectoryListingContract<Driver, M>
+
+    /**
+     * Do recursive listing of items. Without the next function it will do listing of leaf nodes only.
+     * For advanced usage you can pass the next function which will get as parameter current item and it should
+     * return the next location for list or null if the recursion should stop and yield the current item.
+     * For advanced usage you can also limit the depth of recursion using the second argument of next function.
+     */
+    recursive(
+      next?: (current: T, depth: number, driver: Driver) => Promise<string | null> | string | null
+    ): DirectoryListingContract<Driver, T>
+
```

Comment:
Do we really need these additional functions?

Reply:
if you mean `filter` and `map` it is good for developers which prefer functional programming with higher-order functions over imperative style using for loop.
If you mean `pipe` it is added so you can add transforming functions - when extending listing with custom method.